File: shared/types/widgetSnapshot.ts
export type WidgetTaskItem = {
  id: string;
  title: string;
};

/** JSON payload shared between the app and home screen widgets. */
export type WidgetSnapshot = {
  streak: number;
  completedToday: boolean;
  openTaskCount: number;
  tasks: WidgetTaskItem[];
  updatedAt: string;
};

export const WIDGET_SNAPSHOT_STORAGE_KEY = '@acts/widget_snapshot_v1';
export const WIDGET_IOS_DATA_KEY = 'ActsWidgetData';
export const WIDGET_IOS_APP_GROUP = 'group.com.FrogCOO.Acts.expowidgets';
export const ANDROID_WIDGET_STREAK = 'ActsStreakWidget';
export const ANDROID_WIDGET_TASKS = 'ActsTasksWidget';

/** Opens the Tasks tab; optional `taskId` scrolls to that act in the list. */
export function widgetTasksDeepLink(taskId?: string): string {
  const base = 'acts:///(app)/(tabs)/tasks';
  const id = taskId?.trim();
  if (!id) {
    return base;
  }
  return `${base}?taskId=${encodeURIComponent(id)}`;
}

/** Default widget tap target (Tasks tab, no specific act). */
export const WIDGET_TASKS_DEEP_LINK = widgetTasksDeepLink();

File: widgets/ios/WidgetSnapshot.swift
import Foundation

struct WidgetSnapshot: Codable {
    struct TaskItem: Codable {
        let id: String
        let title: String
    }

    let streak: Int
    let completedToday: Bool
    let openTaskCount: Int
    let tasks: [TaskItem]
    let updatedAt: String
}

enum WidgetDataStore {
    static let suiteName = "group.com.FrogCOO.Acts.expowidgets"
    static let dataKey = "ActsWidgetData"

    static func load() -> WidgetSnapshot? {
        guard
            let defaults = UserDefaults(suiteName: suiteName),
            let json = defaults.string(forKey: dataKey),
            let data = json.data(using: .utf8)
        else {
            return nil
        }
        return try? JSONDecoder().decode(WidgetSnapshot.self, from: data)
    }

    static func placeholder() -> WidgetSnapshot {
        WidgetSnapshot(
            streak: 0,
            completedToday: false,
            openTaskCount: 0,
            tasks: [],
            updatedAt: ISO8601DateFormatter().string(from: Date())
        )
    }
}
